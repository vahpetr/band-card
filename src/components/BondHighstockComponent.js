// @ts-check

import React from "react";
import * as Highcharts from "highcharts/highstock";

export class BondHighstockComponent extends React.Component {
  element = null;
  chart = null;

  render() {
    return (
      <div ref={element => (this.element = element)} className="highstock" />
    );
  }

  componentDidMount() {
    this.createChart(this.props);
  }

  componentWillUnmount() {
    this.disposeChart();
  }

  componentWillReceiveProps(nextProps) {
    if (this.chart) {
      this.updateChart(nextProps);
    } else if (this.element) {
      this.createChart(nextProps);
    }
  }

  onValueType(ev) {
    this.props.onFilterChange({
      valueType: ev.target.value
    });
  }

  createChart(options) {
    const config = this.createConfig(options);
    // @ts-ignore
    this.chart = Highcharts.stockChart(this.element, config);
    window.setTimeout(() => {
      this.chart.rangeSelector.clickButton(1, true);
    }, 10);
  }

  updateChart(options) {
    const config = this.createConfig(options);
    this.chart.update(config, true);
  }

  disposeChart() {
    this.chart.destroy();
    this.chart = null;
  }

  createConfig(options) {
    const date = new Date();
    const dateFormat = new Intl.DateTimeFormat("ru").format(date);
    const subtitle =
      `${options.filter.isin}` +
      "<br>" +
      `NII CAPITAL CORP, Telcommunicaions, NR, till ${dateFormat}`;

    const config = {
      chart: {
        type: "line"
      },
      title: {
        text: "NII CAPITAL 7.625 21 USD",
        align: "left"
      },
      subtitle: {
        text: subtitle,
        align: "left"
      },
      credits: {
        enabled: false
      },
      rangeSelector: {
        allButtonsEnabled: true,
        buttonTheme: {
          width: 60
        },
        labelStyle: {
          display: "none"
        },
        inputEnabled: false,
        buttons: [
          {
            count: 1,
            text: "Week",
            type: "week"
          },
          {
            count: 1,
            text: "Month",
            type: "month"
          },
          {
            count: 3,
            text: "Quarter",
            type: "month"
          },
          {
            count: 1,
            text: "Year",
            type: "year"
          },
          {
            text: "Max",
            type: "all"
          }
        ]
      },
      tooltip: {
        shared: true,
        crosshairs: true
      },
      series: [
        {
          name: `${options.filter.isin}`,
          pointStart: Date.UTC(2016, 0, 1),
          pointInterval: 3 * 24 * 3600 * 1000,
          data: options.data,
          marker: {
            enabled: null,
            radius: 2,
            lineWidth: 3,
            lineColor: "#d0d0d0"
          },
          color: "#d0d0d0",
          tooltip: {
            valueDecimals: 3
          }
        }
      ],
      plotOptions: {
        line: {
          dataLabels: {
            enabled: true,
            formatter: function() {
              return Highcharts.numberFormat(this.y, 2);
            }
          }
        }
      },
      xAxis: {
        type: "datetime",
        scrollbar: {
          enabled: false
        },
        tickInterval: 7 * 24 * 3600 * 1000,
        gridLineWidth: 1,
        labels: {
          format: "{value:%d.%m}"
        }
      },
      yAxis: {
        opposite: false
      },

      navigator: {
        enabled: false,
        height: 0,
        margin: 0
      },
      scrollbar: {
        enabled: false
      }
    };
    return config;
  }
}
