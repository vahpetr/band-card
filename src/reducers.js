// @ts-check

import { combineReducers } from "redux";
import {
  BOND_DATA_FETCHING,
  BOND_DATA_SUCCEED,
  BOND_DATA_FAILED
} from "./actions";
import { ValueType } from "./types";

/**
 * App initial state
 */
export const initialState = Object.freeze({
  bond: {
    filter: {
      isin: "RU000A0JU4L3",
      valueType: ValueType.PRICE
    },
    data: [],
    error: null
  }
});

/**
 * Card reducer
 */
export function cardReducer(state = initialState, action) {
  switch (action.type) {
    case BOND_DATA_FETCHING: {
      const newState = { ...state };
      const newFilter = { ...newState.bond.filter, ...action.payload };
      newState.bond.filter = newFilter;
      return newState;
    }
    case BOND_DATA_SUCCEED: {
      const newState = { ...state };
      newState.bond.data = action.payload.data;
      newState.bond.error = null;
      return newState;
    }
    case BOND_DATA_FAILED: {
      const newState = { ...state };
      newState.bond.data = [];
      newState.bond.error = action.payload;
      return newState;
    }
    default: {
      return state;
    }
  }
}

/**
 * App reducer. All combinded reducers
 */
export const appReducers = combineReducers({
  cardReducer
});
