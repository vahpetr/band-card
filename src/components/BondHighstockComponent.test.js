import React from "react";
import { configure, mount } from "enzyme";
import sinon from 'sinon';
import { shallowToJson } from "enzyme-to-json";
import Adapter from "enzyme-adapter-react-16";
import { initialState } from "../reducers";
import { BondHighstockComponent } from "./BondHighstockComponent";

configure({ adapter: new Adapter() });

describe("<BondHighstockComponent />", () => {
  it("renders without crashing", () => {
    const { data, filter } = initialState.bond;
    const wrapper = mount(<BondHighstockComponent data={data} filter={filter} />);

    expect(wrapper.props().filter.isin).toBe(filter.isin);
    expect(wrapper.props().filter.valueType).toBe(filter.valueType);
    expect(wrapper.props().data).toBe(data);
  });

  it("snapshot", () => {
    const { data, filter } = initialState.bond;
    const wrapper = mount(<BondHighstockComponent data={data} filter={filter} />);
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('calls componentDidMount', () => {
    sinon.spy(BondHighstockComponent.prototype, 'componentDidMount');
    const { data, filter } = initialState.bond;
    const wrapper = mount(<BondHighstockComponent data={data} filter={filter} />);
    expect(BondHighstockComponent.prototype.componentDidMount.callCount).toBe(1);
    BondHighstockComponent.prototype.componentDidMount.restore();
  });

});