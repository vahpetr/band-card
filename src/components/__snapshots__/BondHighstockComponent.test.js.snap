// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BondHighstockComponent /> snapshot 1`] = `
<BondHighstockComponent
  data={Array []}
  filter={
    Object {
      "isin": "RU000A0JU4L3",
      "valueType": "Price",
    }
  }
>
  <div
    className="highstock"
  />
</BondHighstockComponent>
`;
